### PR TITLE
Fire VAR events and test.

### PR DESCRIPTION
**Pull request by marisaleung**
All Var Events are fired in locations indicated by the Design Doc:
https://docs.google.com/document/d/1q49IrmDHu-D1ENA5t_Ruxkuej1O2htW5YeM-4Em88ic/edit?ts=592f2c6e&actionButton=1#heading=h.gknl12ppwima

Includes tests for redo and undos of:
- VAR_CREATE
- VAR_DELETE: Without blocks and with blocks
- VAR_RENAME: Without blocks and with blocks

---
This change is [Reviewable](https://reviewable.io/reviews/google/blockly/1162)

### CODE REVIEW

File: core/variable_map.js [hunk 1]
```diff
@@ -81,15 +81,22 @@ Blockly.VariableMap.prototype.renameVariable = function(variable, newName) {
   }
 
   if (variableIndex == -1 && newVariableIndex == -1) {
-    this.createVariable(newName, '');
+    var newVar = this.createVariable(newName, '');
+    Blockly.Events.fire(new Blockly.Events.VarCreate(newVar));
```
Comment (RoboErikG) [inline core/variable_map.js]: Does this lead to two VarCreate events being fired?

Comment (marisaleung) [inline core/variable_map.js]: Yes. Removed!

File: core/variable_map.js [hunk 2]
```diff
@@ -81,15 +81,21 @@ Blockly.VariableMap.prototype.renameVariable = function(variable, newName) {
...
     // Only changing case, or renaming to a completely novel name.
-    this.variableMap_[type][variableIndex].name = newName;
+    var renameVar = this.variableMap_[type][variableIndex];
+    Blockly.Events.fire(new Blockly.Events.VarRename(renameVar, newName));
+    renameVar.name = newName;
   } else if (variableIndex != -1 && newVariableIndex != -1) {
     // Renaming one existing variable to another existing variable.
     // The case might have changed, so we update the destination ID.
+    Blockly.Events.fire(new Blockly.Events.VarRename(
+      this.variableMap_[type][newVariableIndex], newName));
```
Comment (rachel-fenichel) [inline core/variable_map.js]: Stash `this.variableMap_[type][newVariableIndex] ` in a variable--you're using it again on line 98, and it's carrying a lot of information to keep inline.

File: core/variable_map.js [hunk 3]
```diff
@@ -81,15 +81,21 @@ Blockly.VariableMap.prototype.renameVariable = function(variable, newName) {
   }
 
   if (variableIndex == -1 && newVariableIndex == -1) {
-    this.createVariable(newName, '');
+    var newVar = this.createVariable(newName, '');
```
Comment (rachel-fenichel) [inline core/variable_map.js]: Unused variable?

File: core/variable_map.js [hunk 4]
```diff
@@ -81,15 +81,21 @@ Blockly.VariableMap.prototype.renameVariable = function(variable, newName) {
...
     console.log('Tried to rename an non-existent variable.');
   } else if (variableIndex == newVariableIndex ||
       variableIndex != -1 && newVariableIndex == -1) {
     // Only changing case, or renaming to a completely novel name.
-    this.variableMap_[type][variableIndex].name = newName;
+    var renameVar = this.variableMap_[type][variableIndex];
```
Comment (rachel-fenichel) [inline core/variable_map.js]: I'd prefer `variableToRename` and `variableToDelete`, to make it clearer that you are
1) finding the variable
2) firing an event
3) taking the named action on that variable

File: core/variable_map.js (hunk 3 above)
Comment (marisaleung) [inline core/variable_map.js]: Yes, thank you.
Done

File: core/variable_map.js (hunk 4 above)
Comment (marisaleung) [inline core/variable_map.js]: Done.

File: core/variable_map.js (hunk 2 above)
Comment (marisaleung) [inline core/variable_map.js]: Done.

File: tests/jsunit/workspace_undo_redo_test.js [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+  Blockly.Events.fire = temporary_fireEvent;
+}
+
+function undoRedoTest_setUpWithMockBlocks() {
+  undoRedoTest_setUp();
+  // Need to define this because field_variable's dropdownCreate() calls replace
+  // on undefined value, Blockly.Msg.DELETE_VARIABLE. To fix this, define
+  // Blockly.Msg.DELETE_VARIABLE as %1 so the replace function finds the %1 it
+  // expects.
+  Blockly.Msg.DELETE_VARIABLE = '%1';
```
Comment (rachel-fenichel) [inline tests/jsunit/workspace_undo_redo_test.js]: It occurs to me that we can make dropdownCreate check for the existence of Blockly.Msg.DELETE_VARIABLE before trying to call replace on it.

If the message doesn't exist it would just not try to add that option.
Then we could avoid having this hack in so many places.

File: tests/jsunit/workspace_undo_redo_test.js [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+  undoRedoTest_tearDown();
+  Blockly.Msg.DELETE_VARIABLE = saved_msg;
+}
+
+/**
+ * Create a test get_var_block.
+ * @param {string} variable_name The string to put into the variable field.
+ * @return {!Blockly.Block} The created block.
+ */
+function createMockBlock(variable_name) {
```
Comment (rachel-fenichel) [inline tests/jsunit/workspace_undo_redo_test.js]: This name is fine in isolation, but may be problematic because all of the test files are loaded in the same namespace.  Prefix it with something that makes it clear it belongs to this file, or move it to test_utilities if it's going to be defined or used in multiple test files.

As mentioned below, if it's going into test_utilities you can do it in a followup PR and leave it as-is for this PR.

File: tests/jsunit/workspace_undo_redo_test.js [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+'use strict';
+
+goog.require('goog.events.EventHandler');
+goog.require('goog.testing');
+goog.require('goog.testing.events');
+goog.require('goog.testing.MockControl');
+
+var workspace;
+var mockControl_;
+var saved_msg = Blockly.Msg.DELETE_VARIABLE;
```
Comment (rachel-fenichel) [inline tests/jsunit/workspace_undo_redo_test.js]: Use camelCase consistently, instead of switching to snake_case for some variables.

File: tests/jsunit/workspace_undo_redo_test.js [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+  return block;
+}
+
+/**
+ * Check if a variable with the given values exists.
+ * @param {string} name The expected name of the variable.
+ * @param {string} type The expected type of the variable.
+ * @param {string} id The expected id of the variable.
+ */
+function undoRedoTest_checkVariableValues(name, type, id) {
```
Comment (rachel-fenichel) [inline tests/jsunit/workspace_undo_redo_test.js]: You've defined a similar function in workspace_test.js, xml_test.js, and variable_map_test.js.  You should combine them into a single function in tests/test_utilities.js.

In the interest of keeping this PR clean, leave this function here for now.  Then make a separate PR that just moves it to test_utilities and removes duplicates from your other test files.

File: tests/jsunit/workspace_undo_redo_test.js [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+  var variable = workspace.getVariable(name);
+  assertNotUndefined(variable);
+  assertEquals(name, variable.name);
+  assertEquals(type, variable.type);
+  assertEquals(id, variable.getId());
+}
+
+/**
+ * Check if a variable names in specified top block.
```
Comment (rachel-fenichel) [inline tests/jsunit/workspace_undo_redo_test.js]: Can't parse.

File: tests/jsunit/workspace_undo_redo_test.js [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+  undoRedoTest_tearDown();
+}
+
+function test_undoDeleteVariable_WithBlocks() {
+  undoRedoTest_setUpWithMockBlocks();
+  workspace.createVariable('name1', 'type1', 'id1');
+  workspace.createVariable('name2', 'type2', 'id2');
+  createMockBlock('name1');
+  createMockBlock('name2');
+  workspace.deleteVariableById('id1');
```
Comment (rachel-fenichel) [inline tests/jsunit/workspace_undo_redo_test.js]: It occurs to me that you should test what happens if you delete a variable by id twice--that is, test deleting a variable that doesn't exist.  Again, doesn't have  to be in this PR.

File: tests/jsunit/workspace_undo_redo_test.js [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+  workspace.undo();
+  workspace.undo(true);
+  // Expect that variable 'id2' is recreated
+  undoRedoTest_checkBlockVariableName(0, 'name2');
+  assertNull(workspace.getVariableById('id1'));
+  undoRedoTest_checkVariableValues('name2', 'type2', 'id2');
+  undoRedoTest_tearDownWithMockBlocks();
+}
+
+function test_undoRedoRenameVariable_NeitherVariableExists() {
```
Comment (rachel-fenichel) [inline tests/jsunit/workspace_undo_redo_test.js]: Can you explain this test to me?

File: tests/jsunit/workspace_undo_redo_test.js (hunk 7 above)
Comment (marisaleung) [inline tests/jsunit/workspace_undo_redo_test.js]: Done, changed a couple of other variables in the file to use camelCase instead of snake.

File: tests/jsunit/workspace_undo_redo_test.js (hunk 6 above)
Comment (marisaleung) [inline tests/jsunit/workspace_undo_redo_test.js]: Will do in a separate CL.

File: tests/jsunit/workspace_undo_redo_test.js (hunk 8 above)
Comment (marisaleung) [inline tests/jsunit/workspace_undo_redo_test.js]: Will do in a separate PR.

File: tests/jsunit/workspace_undo_redo_test.js (hunk 10 above)
Comment (marisaleung) [inline tests/jsunit/workspace_undo_redo_test.js]: Good idea, will do in separate PR.

File: tests/jsunit/workspace_undo_redo_test.js (hunk 11 above)
Comment (marisaleung) [inline tests/jsunit/workspace_undo_redo_test.js]: Added an explaining comment. For reference it is:
"Expect that a variable with the name, 'name2', and the generated UUID, 'id2', to be created when rename is called. Undo removes this variable and redo recreates it."

File: tests/jsunit/workspace_undo_redo_test.js (hunk 5 above)
Comment (marisaleung) [inline tests/jsunit/workspace_undo_redo_test.js]: I started making this change and realized a lot of test files get to be shortened by doing this. So I'll do this in a separate PR as well.

Comment (rachel-fenichel) [inline tests/jsunit/workspace_undo_redo_test.js]: SGTM.  I think most of my comments on this PR were requests to do completely separate cleanup PRs :)

Comment (marisaleung) [inline tests/jsunit/workspace_undo_redo_test.js]: Done
